File: noop-upstream/Strand/Data/BackupSync.swift
import Foundation
#if os(macOS)
import AppKit
#else
import UIKit
import UniformTypeIdentifiers
#endif

/// Backup & Sync (folder destination) - the Apple twin of the Android `BackupSync`. Writes the full
/// `.noopbak` snapshot (the existing `DataBackup` format) into a user-chosen folder, on demand and as
/// an on-launch daily catch-up. Point that folder at a Google Drive / iCloud / Dropbox client and you
/// get automatic off-device backup with no in-app cloud account - NOOP only writes a local file; the
/// sync client does the upload.
///
/// `BackupSync` holds only the PURE, unit-tested filename / selection logic (no I/O, no state) so it is
/// byte-for-byte equivalent to the Android twin: same `noop-backup-YYYYMMDD-HHMMSS.noopbak` scheme,
/// same newest-first selection, same keep-N prune. The stateful folder I/O lives in `FolderBackup`.
enum BackupSync {

    /// The destinations this is built to support. Phase 1 ships `.folder`; a future Drive backend slots
    /// in behind the same UI/logic without changing the snapshot scheme.
    enum Destination { case folder /* , googleDrive */ }

    static let prefix = "noop-backup-"
    static let suffix = ".noopbak"

    // MARK: - Pure helpers (unit-tested; mirror Android BackupSync)

    /// A fresh UTC second-resolution formatter. Created per call so the type stays trivially `Sendable`
    /// and there is no shared mutable `DateFormatter` to data-race; the cost is irrelevant at this rate.
    private static func stampFormatter() -> DateFormatter {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(identifier: "UTC")
        f.dateFormat = "yyyyMMdd-HHmmss"
        f.isLenient = false
        return f
    }

    /// Canonical snapshot filename for an instant (ms since epoch): `noop-backup-YYYYMMDD-HHMMSS.noopbak` (UTC).
    static func snapshotName(_ epochMs: Int) -> String {
        prefix + stampFormatter().string(from: Date(timeIntervalSince1970: Double(epochMs) / 1000.0)) + suffix
    }

    /// The UTC instant (ms) encoded in a snapshot filename, or nil if `name` is not one of ours.
    static func snapshotTimeMs(_ name: String) -> Int? {
        guard name.hasPrefix(prefix), name.hasSuffix(suffix) else { return nil }
        let stamp = String(name.dropFirst(prefix.count).dropLast(suffix.count))
        guard let d = stampFormatter().date(from: stamp) else { return nil }
        return Int((d.timeIntervalSince1970 * 1000.0).rounded())
    }

    static func isSnapshot(_ name: String) -> Bool { snapshotTimeMs(name) != nil }

    /// Newest snapshot by encoded time (non-snapshots ignored), or nil.
    static func latestSnapshot(_ names: [String]) -> String? {
        names.filter(isSnapshot).max { (snapshotTimeMs($0) ?? 0) < (snapshotTimeMs($1) ?? 0) }
    }

    /// Snapshot names sorted newest-first (non-snapshots dropped). The folder restore list order.
    static func snapshotsNewestFirst(_ names: [String]) -> [String] {
        names.filter(isSnapshot).sorted { (snapshotTimeMs($0) ?? 0) > (snapshotTimeMs($1) ?? 0) }
    }

    /// Snapshots to DELETE to keep only the `keep` newest (oldest-first). Empty when within budget.
    static func snapshotsToPrune(_ names: [String], keep: Int) -> [String] {
        let snaps = snapshotsNewestFirst(names)
        return snaps.count <= keep ? [] : Array(snaps.dropFirst(keep))
    }
}

/// The folder destination: a security-scoped bookmark of a user-chosen folder, plus write / list /
/// restore / prune and an on-launch daily catch-up.
///
/// Apple gives no reliable unattended background DB write (a whole-DB ZIP can be 100 MB+ and the OS
/// kills long background tasks), so the "schedule" is a deferred on-launch catch-up: if auto is ON and
/// a day has passed, the next foreground launch writes one snapshot - fully off the main thread, and
/// off the launch-critical path, so it never blocks startup (must-fix #4).
enum FolderBackup {
    private static let bookmarkKey = "backupSync.folderBookmark"
    private static let autoKey = "backupSync.auto"
    private static let lastKey = "backupSync.lastMs"

    /// Keep the latest N snapshots in the folder; older ones are pruned. Matches the Android default.
    static let keepCount = 10
    private static let dayMs = 24 * 60 * 60 * 1000

    // MARK: - Persisted state

    /// Auto-backup defaults OFF (manual-first; must-fix #4 + #6). The user turns it on in the screen.
    static var autoEnabled: Bool {
        get { UserDefaults.standard.bool(forKey: autoKey) }
        set { UserDefaults.standard.set(newValue, forKey: autoKey) }
    }

    static var lastBackupMs: Int { UserDefaults.standard.integer(forKey: lastKey) }
    static var hasFolder: Bool { UserDefaults.standard.data(forKey: bookmarkKey) != nil }

    /// A short, human label for the chosen folder (its last path component), or nil if none chosen.
    static func folderLabel() -> String? { resolveFolder()?.lastPathComponent }

    // MARK: - Security-scoped bookmark

    private static func bookmarkCreationOptions() -> URL.BookmarkCreationOptions {
        #if os(macOS)
        return [.withSecurityScope]
        #else
        return []
        #endif
    }

    private static func resolveFolder() -> URL? {
        guard let data = UserDefaults.standard.data(forKey: bookmarkKey) else { return nil }
        var stale = false
        #if os(macOS)
        let opts: URL.BookmarkResolutionOptions = [.withSecurityScope]
        #else
        let opts: URL.BookmarkResolutionOptions = []
        #endif
        let url = try? URL(resolvingBookmarkData: data, options: opts, relativeTo: nil, bookmarkDataIsStale: &stale)
        // A stale bookmark still resolves; refresh it opportunistically so it keeps working next launch.
        if stale, let url { saveFolder(url) }
        return url
    }

    /// Persist a security-scoped bookmark for `url`. Brackets the bookmark creation in a scoped-access
    /// pair on macOS (required to mint a `.withSecurityScope` bookmark for a sandbox-granted folder).
    static func saveFolder(_ url: URL) {
        let scoped = url.startAccessingSecurityScopedResource()
        defer { if scoped { url.stopAccessingSecurityScopedResource() } }
        if let data = try? url.bookmarkData(options: bookmarkCreationOptions(),
                                            includingResourceValuesForKeys: nil, relativeTo: nil) {
            UserDefaults.standard.set(data, forKey: bookmarkKey)
        }
    }

    // MARK: - Backup / prune

    /// Write one snapshot into the bookmarked folder, stamp the last-backup time, then prune to keepN.
    /// Returns true on success. Runs the whole-DB ZIP off the caller's thread; never touches UI.
    @discardableResult
    static func backupNow(checkpoint: @escaping () async -> Bool) async -> Bool {
        guard let folder = resolveFolder() else { return false }
        let scoped = folder.startAccessingSecurityScopedResource()
        defer { if scoped { folder.stopAccessingSecurityScopedResource() } }

        let nowMs = Int(Date().timeIntervalSince1970 * 1000.0)
        let dest = folder.appendingPathComponent(BackupSync.snapshotName(nowMs))
        guard case .exported = await DataBackup.writeBackup(checkpoint: checkpoint, to: dest) else { return false }

        UserDefaults.standard.set(nowMs, forKey: lastKey)
        prune(in: folder)
        return true
    }

    /// Best-effort retention: delete snapshots beyond `keepCount`. Listing failures are ignored, and
    /// non-snapshot files in the folder are never touched (only files matching the backup scheme).
    private static func prune(in folder: URL) {
        let names = (try? FileManager.default.contentsOfDirectory(atPath: folder.path)) ?? []
        let toDelete = Set(BackupSync.snapshotsToPrune(names, keep: keepCount))
        for name in names where toDelete.contains(name) {
            try? FileManager.default.removeItem(at: folder.appendingPathComponent(name))
        }
    }

    /// On-launch catch-up: if auto is on, a folder is set, and it's been at least a day since the last
    /// backup, write one. Gated entirely on the toggle being ON (must-fix #4). The caller MUST invoke
    /// this off the launch-critical path; it does no UI work and is safe to run after the first refresh.
    static func catchUpIfDue(checkpoint: @escaping () async -> Bool) async {
        guard autoEnabled, hasFolder else { return }
        let nowMs = Int(Date().timeIntervalSince1970 * 1000.0)
        guard nowMs - lastBackupMs >= dayMs else { return }
        await backupNow(checkpoint: checkpoint)
    }

    // MARK: - Restore from the configured folder (must-fix #1)

    /// One restorable snapshot in the chosen folder: its display name + a friendly absolute-time line.
    struct Snapshot: Identifiable, Hashable {
        let name: String
        let timeMs: Int
        var id: String { name }
    }

    /// List the snapshots in the user's chosen backup folder, NEWEST FIRST - the source for the restore
    /// picker (must-fix #1: restore from the configured folder, never a generic re-prompt). Returns an
    /// empty list if no folder is chosen or the folder can't be read.
    static func listSnapshots() -> [Snapshot] {
        guard let folder = resolveFolder() else { return [] }
        let scoped = folder.startAccessingSecurityScopedResource()
        defer { if scoped { folder.stopAccessingSecurityScopedResource() } }
        let names = (try? FileManager.default.contentsOfDirectory(atPath: folder.path)) ?? []
        return BackupSync.snapshotsNewestFirst(names).compactMap { name in
            guard let ms = BackupSync.snapshotTimeMs(name) else { return nil }
            return Snapshot(name: name, timeMs: ms)
        }
    }

    /// Restore a snapshot the user picked FROM THE CONFIGURED FOLDER. Resolves the bookmark, brackets
    /// security-scoped access, and hands the concrete file to `DataBackup.restore(from:)` - so every
    /// hardened safety (magic-byte + GRDB-origin validation, sidecar snapshot, rollback) still applies.
    /// The destructive confirmation is the screen's responsibility (must-fix #2); this is the final
    /// step it calls only AFTER the user confirms.
    static func restore(snapshotNamed name: String) -> DataBackup.BackupResult {
        guard let folder = resolveFolder() else {
            return .failure("Couldn't open your backup folder - re-pick it and try again.")
        }
        let scoped = folder.startAccessingSecurityScopedResource()
        defer { if scoped { folder.stopAccessingSecurityScopedResource() } }
        let source = folder.appendingPathComponent(name)
        guard FileManager.default.fileExists(atPath: source.path) else {
            return .failure("That backup is no longer in your folder.")
        }
        return DataBackup.restore(from: source)
    }

    // MARK: - Folder picker

    #if os(macOS)
    /// Present an `NSOpenPanel` to choose a directory; persists the bookmark. Returns the chosen URL.
    @MainActor
    static func pickFolder() -> URL? {
        let panel = NSOpenPanel()
        panel.canChooseDirectories = true
        panel.canChooseFiles = false
        panel.allowsMultipleSelection = false
        panel.prompt = "Choose"
        panel.message = "Choose a folder for NOOP backups (for example a Google Drive or iCloud folder)."
        guard panel.runModal() == .OK, let url = panel.url else { return nil }
        saveFolder(url)
        return url
    }
    #else
    /// Present a folder picker (`UIDocumentPicker`) and persist the bookmark. Returns the chosen URL.
    @MainActor
    static func pickFolder() async -> URL? {
        let url = await DocumentPicker.pickFolder()
        if let url { saveFolder(url) }
        return url
    }
    #endif
}

File: noop-upstream/android/app/src/main/java/com/noop/ui/BackupSync.kt
package com.noop.ui

import android.content.Context
import android.net.Uri
import android.provider.DocumentsContract
import androidx.work.CoroutineWorker
import androidx.work.ExistingPeriodicWorkPolicy
import androidx.work.PeriodicWorkRequestBuilder
import androidx.work.WorkManager
import androidx.work.WorkerParameters
import com.noop.data.DataBackup
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.TimeUnit

/**
 * Backup & Sync (Phase 1 - folder destination). Writes the full `.noopbak` snapshot (the existing
 * [DataBackup] whole-DB format) into a user-chosen folder (a SAF tree), on demand and on an opt-in
 * daily schedule. Point that folder at a desktop Google Drive / Dropbox sync client (or a phone sync
 * app) and you get off-device backup with NO in-app cloud account, no OAuth, no secrets - NOOP only
 * ever writes a local file; the user's own sync client does any upload.
 *
 * DESIGN
 * - Snapshots are timestamped and immutable. "Restore" REPLACES the live DB (whole-DB snapshot,
 *   newest-wins), exactly as [DataBackup.importFrom] already does - we add nothing to the restore
 *   safety path (magic-byte + Room/GRDB-origin validation, sidecar snapshot, rollback-on-failure).
 * - The pure filename/selection helpers are unit-tested byte-for-byte against the Apple twin so a
 *   `.noopbak` produced on either platform is named + selected identically.
 * - The daily schedule is opt-in (default OFF) and runs off the main thread (a whole-DB zip can be
 *   100MB+). The periodic write goes through WorkManager (survives reboot/app-kill, never the
 *   launch-critical path); the on-launch CATCH-UP is a deferred IO coroutine (see [catchUpIfDue]).
 */
object BackupSync {

    /** The two backup destinations Phase 1 is built to support (Drive lands in a later phase). */
    enum class Destination { FOLDER /* , GOOGLE_DRIVE */ }

    private const val PREFIX = "noop-backup-"
    private const val SUFFIX = ".noopbak"

    /** Generic binary MIME for the SAF createDocument call (the bytes are a ZIP container). */
    const val MIME = "application/octet-stream"

    /** Default snapshots kept by prune. Identical to the Apple `FolderBackup.keepCount`. */
    const val DEFAULT_KEEP = 10

    /** A day in ms - the catch-up cadence (mirrors the Apple `dayMs`). */
    private const val DAY_MS = 24L * 60L * 60L * 1000L

    // ── Pure helpers (unit-tested; mirror the Apple BackupSync) ─────────────────

    private fun fmt() = SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
        isLenient = false
    }

    /** Canonical snapshot filename for an instant: `noop-backup-YYYYMMDD-HHMMSS.noopbak` (UTC). */
    fun snapshotName(epochMs: Long): String = PREFIX + fmt().format(Date(epochMs)) + SUFFIX

    /** The UTC instant (ms) encoded in a snapshot filename, or null if [name] is not one of ours. */
    fun snapshotTimeMs(name: String): Long? {
        if (!name.startsWith(PREFIX) || !name.endsWith(SUFFIX)) return null
        val stamp = name.substring(PREFIX.length, name.length - SUFFIX.length)
        return runCatching { fmt().parse(stamp)?.time }.getOrNull()
    }

    fun isSnapshot(name: String): Boolean = snapshotTimeMs(name) != null

    /** Newest snapshot by encoded time among [names] (non-snapshots ignored), or null if none. */
    fun latestSnapshot(names: List<String>): String? =
        names.filter(::isSnapshot).maxByOrNull { snapshotTimeMs(it)!! }

    /** Snapshots to DELETE to keep only the [keep] newest (oldest-first). Empty when within budget. */
    fun snapshotsToPrune(names: List<String>, keep: Int): List<String> {
        val snaps = names.filter(::isSnapshot).sortedByDescending { snapshotTimeMs(it)!! }
        return if (snaps.size <= keep) emptyList() else snaps.drop(keep)
    }

    /** True when [nowMs] is at least a day past [lastBackupMs] (pure, so the catch-up gate is testable). */
    fun isCatchUpDue(lastBackupMs: Long, nowMs: Long): Boolean = nowMs - lastBackupMs >= DAY_MS

    // ── Folder destination I/O (SAF tree via DocumentsContract - no extra dep) ──

    /** Create + write one snapshot into the chosen [treeUri]; returns the new file Uri, or null on failure. */
    fun writeSnapshot(context: Context, treeUri: Uri, nowMs: Long = System.currentTimeMillis()): Uri? {
        val resolver = context.contentResolver
        val parentDoc = DocumentsContract.buildDocumentUriUsingTree(
            treeUri,
            DocumentsContract.getTreeDocumentId(treeUri),
        )
        val fileUri = runCatching {
            DocumentsContract.createDocument(resolver, parentDoc, MIME, snapshotName(nowMs))
        }.getOrNull() ?: return null
        // exportTo throws on failure; on a partial write delete the half-written doc so prune/latest
        // never picks up a corrupt snapshot.
        return runCatching {
            DataBackup.exportTo(context, fileUri)
            fileUri
        }.getOrElse {
            runCatching { DocumentsContract.deleteDocument(resolver, fileUri) }
            null
        }
    }

    /** Run one backup into the persisted folder, prune to [BackupSyncPrefs.keepCount], stamp last-backup. */
    fun backupNow(context: Context): Boolean {
        val treeUri = BackupSyncPrefs.treeUri(context) ?: return false
        writeSnapshot(context, treeUri) ?: return false
        BackupSyncPrefs.setLastBackupMs(context, System.currentTimeMillis())
        prune(context, treeUri)
        return true
    }

    /** Best-effort retention: delete snapshots beyond keepCount. Listing failures are ignored. */
    private fun prune(context: Context, treeUri: Uri) {
        val keep = BackupSyncPrefs.keepCount(context)
        val children = runCatching { listSnapshotDocs(context, treeUri) }.getOrDefault(emptyList())
        val toDelete = snapshotsToPrune(children.map { it.first }, keep).toSet()
        for ((name, docUri) in children) {
            if (name in toDelete) {
                runCatching { DocumentsContract.deleteDocument(context.contentResolver, docUri) }
            }
        }
    }

    /** (name, documentUri) for every snapshot in the tree, newest-first. Drives restore-from-folder. */
    fun listSnapshotDocs(context: Context, treeUri: Uri): List<Pair<String, Uri>> {
        val childrenUri = DocumentsContract.buildChildDocumentsUriUsingTree(
            treeUri,
            DocumentsContract.getTreeDocumentId(treeUri),
        )
        val out = ArrayList<Pair<String, Uri>>()
        context.contentResolver.query(
            childrenUri,
            arrayOf(
                DocumentsContract.Document.COLUMN_DOCUMENT_ID,
                DocumentsContract.Document.COLUMN_DISPLAY_NAME,
            ),
            null, null, null,
        )?.use { c ->
            while (c.moveToNext()) {
                val id = c.getString(0)
                val name = c.getString(1) ?: continue
                if (isSnapshot(name)) {
                    out.add(name to DocumentsContract.buildDocumentUriUsingTree(treeUri, id))
                }
            }
        }
        return out.sortedByDescending { snapshotTimeMs(it.first)!! }
    }

    // ── Scheduling (WorkManager - survives reboot/app-kill, mirrors DebugExportScheduler) ──

    private const val WORK = "noop_backup_sync_daily"

    /**
     * (Re)schedule (or cancel) the daily auto-backup from the persisted state. No-op + cancels when the
     * feature is off or no folder is chosen, so toggling off stops the writes. KEEP so re-enabling or a
     * reboot doesn't stack duplicate daily jobs. Wrap call sites so a WorkManager hiccup never throws.
     */
    fun reschedule(context: Context) {
        val wm = WorkManager.getInstance(context.applicationContext)
        if (!BackupSyncPrefs.autoEnabled(context) || BackupSyncPrefs.treeUri(context) == null) {
            wm.cancelUniqueWork(WORK)
            return
        }
        val req = PeriodicWorkRequestBuilder<BackupSyncWorker>(1, TimeUnit.DAYS).build()
        wm.enqueueUniquePeriodicWork(WORK, ExistingPeriodicWorkPolicy.KEEP, req)
    }

    /**
     * On-launch CATCH-UP backup. Must-fix #4: deferred off the launch-critical path and run fully off
     * the main thread by the CALLER (MainActivity launches this on Dispatchers.IO after the critical
     * startup work). Gated on the toggle being ON, a folder being set, and it being at least a day since
     * the last backup. Cheap to call when off (two SharedPreferences reads, no I/O). Returns true if it
     * actually wrote a backup.
     */
    fun catchUpIfDue(context: Context, nowMs: Long = System.currentTimeMillis()): Boolean {
        if (!BackupSyncPrefs.autoEnabled(context)) return false
        if (BackupSyncPrefs.treeUri(context) == null) return false
        if (!isCatchUpDue(BackupSyncPrefs.lastBackupMs(context), nowMs)) return false
        return backupNow(context)
    }
}

/**
 * The worker that performs one scheduled backup. Re-checks the toggle (a stale enqueued job from before
 * the user turned auto off must not fire) and writes one snapshot into the chosen folder. A transient
 * failure (e.g. the folder's sync app has the tree busy) returns retry rather than poisoning the chain.
 */
class BackupSyncWorker(appContext: Context, params: WorkerParameters) :
    CoroutineWorker(appContext, params) {
    override suspend fun doWork(): Result {
        if (!BackupSyncPrefs.autoEnabled(applicationContext)) return Result.success()
        return if (BackupSync.backupNow(applicationContext)) Result.success() else Result.retry()
    }
}

/**
 * Small SharedPreferences-backed store for the folder destination + schedule state. Mirrors the Apple
 * `FolderBackup` UserDefaults keys: a persisted SAF tree uri, the auto toggle (default OFF), the last
 * backup instant, and the keep-N (default [BackupSync.DEFAULT_KEEP]). Single-user, on-device.
 */
object BackupSyncPrefs {
    private const val FILE = "backup_sync"
    private fun p(c: Context) = c.applicationContext.getSharedPreferences(FILE, Context.MODE_PRIVATE)

    fun treeUri(c: Context): Uri? = p(c).getString("tree_uri", null)?.let(Uri::parse)
    fun setTreeUri(c: Context, uri: Uri?) = p(c).edit().apply {
        if (uri == null) remove("tree_uri") else putString("tree_uri", uri.toString())
    }.apply()

    /** Master enable for the daily auto-backup. Default OFF (every NOOP automation is opt-in). */
    fun autoEnabled(c: Context): Boolean = p(c).getBoolean("auto", false)
    fun setAutoEnabled(c: Context, on: Boolean) = p(c).edit().putBoolean("auto", on).apply()

    fun lastBackupMs(c: Context): Long = p(c).getLong("last_ms", 0L)
    fun setLastBackupMs(c: Context, ms: Long) = p(c).edit().putLong("last_ms", ms).apply()

    fun keepCount(c: Context): Int = p(c).getInt("keep", BackupSync.DEFAULT_KEEP)
    fun setKeepCount(c: Context, n: Int) = p(c).edit().putInt("keep", n.coerceIn(1, 100)).apply()
}
